speechbrain/nnet/RNN.py

```diff
@@ -1269,7 +1292,412 @@ def _ligru_cell(self, w, ht):
+                    )
+                ).data
+
+            # Sampling the mask
+            drop_mask = self.drop_masks[
+                self.drop_mask_cnt : self.drop_mask_cnt + self.batch_size
+            ]
+            self.drop_mask_cnt = self.drop_mask_cnt + self.batch_size
+
+        else:
+            self.drop_mask_te = self.drop_mask_te.to(w.device)
+            drop_mask = self.drop_mask_te
+
+        return drop_mask
+
+    def _change_batch_size(self, x):
+        """This function changes the batch size when it is different from
+        the one detected in the initialization method. This might happen in
+        the case of multi-gpu or when we have different batch sizes in train
+        and test. We also update the h_int and drop masks.
+        """
+        if self.batch_size != x.shape[0]:
+            self.batch_size = x.shape[0]
+
+            if self.training:
+                self.drop_masks = self.drop(
+                    torch.ones(
+                        self.N_drop_masks, self.hidden_size, device=x.device,
+                    )
+                ).data
+
+
+class SLiGRU(torch.nn.Module):
+    """ This class implements a Stabilised Light GRU (Li-GRU).
+
+    SLi-GRU is single-gate GRU model based on batch-norm + relu
+    activations + layer-norm on the recurrent connections + recurrent dropout.
+
+    The SLi-GRU differs from the vanilla Li-GRU on the recurrent weights. Indeed, the Li-GRU
+    suffers from an exploding gradient problem on the recurrent weights, and cannot be trained on medium to large ASR dataset.
+    To solve this problem, we use a layer-norm on the recurrent weights that stabilises the training of the model and allows one
+    to train it on large ASR datasets without any problem.
+
+    This model beat traditional LSTM/GRU models on the CommonVoice/LibriSpeech datasets (WER and efficiency).
+
+    For more info see:
+    "Moumen, Adel and Parcollet, Titouan,
+    Stabilising and accelerating light gated recurrent units for automatic speech recognition,
+    in 2023 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP)"
+    (https://arxiv.org/abs/2302.10144)
+
+    To improve the speed of the model, it is recommended to use the torch just-in-time compiler (jit)
+    right before using it.
+
+    You can compile it with:
+    compiled_model = torch.jit.script(model)
+
+    It accepts in input tensors formatted as (batch, time, fea).
+    In the case of 4d inputs like (batch, time, fea, channel) the tensor is
+    flattened as (batch, time, fea*channel).
+
+    Arguments
+    ---------
+    hidden_size : int
+        Number of output neurons (i.e, the dimensionality of the output).
+        values (i.e, time and frequency kernel sizes respectively).
+    input_shape : tuple
+        The shape of an example input.
+    nonlinearity : str
+        Type of nonlinearity (tanh, relu).
+    ff_normalization : str
```

Comment:
The paper doesn't mention (often?) the feedforward LN/BN so IMO, a reader could easily confuse SLi-GRU's LN in the recurrent layer with the feedforward LN/BN and believe that it refers to the LN. It might be worth mentioning something among the lines of "Note that a LayerNorm is applied in the recurrent network regardless of normalization applied on the feed-forward network"?

Reply:
Yes we could do that.

Reply:
Done